File: mobile-worker/lib/offline-queue.ts
import * as SQLite from 'expo-sqlite';
import { api } from './api';

/**
 * Offline-first sync queue for completed refills, reclaims and walk-in
 * sales. Each operation is written to the local DB *first*, then the
 * uploader drains the queue when connectivity returns. Drivers can keep
 * working even with zero bars of signal — critical in Iraq.
 *
 * The schema is intentionally minimal: one row per pending mutation,
 * with the HTTP method, path, and JSON body. The uploader iterates rows
 * oldest-first and deletes on 2xx response.
 */
let dbInstance: SQLite.SQLiteDatabase | null = null;

async function getDB() {
  if (dbInstance) return dbInstance;
  dbInstance = await SQLite.openDatabaseAsync('maa-worker.db');
  await dbInstance.execAsync(`
    CREATE TABLE IF NOT EXISTS pending_mutations (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      method TEXT NOT NULL,
      path TEXT NOT NULL,
      body TEXT NOT NULL,
      created_at INTEGER NOT NULL,
      retries INTEGER NOT NULL DEFAULT 0,
      last_error TEXT
    );
  `);
  return dbInstance;
}

export async function enqueue(method: 'POST' | 'PATCH', path: string, body: any) {
  const db = await getDB();
  await db.runAsync(
    'INSERT INTO pending_mutations (method, path, body, created_at) VALUES (?, ?, ?, ?)',
    method,
    path,
    JSON.stringify(body),
    Date.now(),
  );
}

export async function pendingCount(): Promise<number> {
  const db = await getDB();
  const row = await db.getFirstAsync<{ n: number }>(
    'SELECT COUNT(*) AS n FROM pending_mutations',
  );
  return row?.n ?? 0;
}

/**
 * Tries to upload everything in the queue. Stops on the first failure
 * so we don't burn battery hammering a dead server. Caller should retry
 * later (e.g. on connectivity-change events or app foreground).
 */
export async function flush(): Promise<{ ok: number; failed: number }> {
  const db = await getDB();
  const rows = await db.getAllAsync<{
    id: number;
    method: string;
    path: string;
    body: string;
    retries: number;
  }>('SELECT id, method, path, body, retries FROM pending_mutations ORDER BY id ASC');

  let ok = 0;
  let failed = 0;
  for (const r of rows) {
    try {
      await api.request({
        method: r.method,
        url: r.path,
        data: JSON.parse(r.body),
      });
      await db.runAsync('DELETE FROM pending_mutations WHERE id = ?', r.id);
      ok++;
    } catch (e: any) {
      failed++;
      await db.runAsync(
        'UPDATE pending_mutations SET retries = retries + 1, last_error = ? WHERE id = ?',
        String(e?.message ?? e),
        r.id,
      );
      // If it's a 4xx (validation error), discard — retrying won't help.
      const status = e?.response?.status ?? 0;
      if (status >= 400 && status < 500 && status !== 401 && status !== 408 && status !== 429) {
        await db.runAsync('DELETE FROM pending_mutations WHERE id = ?', r.id);
      }
      break; // stop on first failure
    }
  }
  return { ok, failed };
}

File: mobile-flutter/packages/daari_core/lib/src/services/offline_queue.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:sqflite/sqflite.dart';

/// طابور الطفرات الأوفلاين للسائق — منقول من `worker/lib/offline-queue.ts`.
///
/// عند فشل الشبكة تُكتب العملية (إكمال طلب / استرجاع خزان / بيع فوري / تسجيل
/// زبون) في جدول SQLite محلي، ثم يصرّفها [flush] عند عودة الاتصال. هكذا يواصل
/// السائق العمل بإشارة صفر — حرجٌ في العراق.
///
/// الصفّ الواحد = طفرة معلّقة (method + path + body JSON). المُصرِّف يمرّ على
/// الصفوف الأقدم أولاً ويحذف عند 2xx، ويُسقِط أخطاء 4xx (عدا 401/408/429) لأن
/// إعادتها لن تنفع، ويتوقّف عند أول فشل شبكة كي لا يستنزف البطارية.
class OfflineQueue {
  OfflineQueue(this._dio);

  final Dio _dio;
  Database? _db;

  Future<Database> _open() async {
    final existing = _db;
    if (existing != null) return existing;
    final dir = await getDatabasesPath();
    final db = await openDatabase(
      '$dir/daari_worker.db',
      version: 1,
      onCreate: (db, _) => db.execute('''
        CREATE TABLE pending_mutations (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          method TEXT NOT NULL,
          path TEXT NOT NULL,
          body TEXT NOT NULL,
          created_at INTEGER NOT NULL,
          retries INTEGER NOT NULL DEFAULT 0,
          last_error TEXT
        )
      '''),
    );
    _db = db;
    return db;
  }

  /// أضِف طفرة معلّقة (تُرسَل لاحقاً عند عودة الشبكة).
  Future<void> enqueue(String method, String path, Map<String, dynamic> body) async {
    final db = await _open();
    await db.insert('pending_mutations', {
      'method': method,
      'path': path,
      'body': jsonEncode(body),
      'created_at': DateTime.now().millisecondsSinceEpoch,
    });
  }

  /// عدد الطفرات المنتظرة المزامنة.
  Future<int> pendingCount() async {
    final db = await _open();
    final rows =
        await db.rawQuery('SELECT COUNT(*) AS n FROM pending_mutations');
    return Sqflite.firstIntValue(rows) ?? 0;
  }

  /// يحاول إرسال كل ما في الطابور (الأقدم أولاً). يتوقّف عند أول فشل شبكة.
  Future<({int ok, int failed})> flush() async {
    final db = await _open();
    final rows = await db.query('pending_mutations', orderBy: 'id ASC');
    var ok = 0;
    var failed = 0;
    for (final row in rows) {
      final id = row['id'] as int;
      try {
        await _dio.request<dynamic>(
          row['path'] as String,
          data: jsonDecode(row['body'] as String),
          options: Options(method: row['method'] as String),
        );
        await db.delete('pending_mutations', where: 'id = ?', whereArgs: [id]);
        ok++;
      } on DioException catch (e) {
        failed++;
        final status = e.response?.statusCode ?? 0;
        await db.update(
          'pending_mutations',
          {
            'retries': ((row['retries'] as int?) ?? 0) + 1,
            'last_error': e.message,
          },
          where: 'id = ?',
          whereArgs: [id],
        );
        // خطأ تحقّق دائم (4xx) لن تُصلحه الإعادة → احذفه. عدا 401/408/429.
        if (status >= 400 &&
            status < 500 &&
            status != 401 &&
            status != 408 &&
            status != 429) {
          await db
              .delete('pending_mutations', where: 'id = ?', whereArgs: [id]);
        }
        break; // توقّف عند أول فشل
      }
    }
    return (ok: ok, failed: failed);
  }
}
